Java/src/com/couchbase/litecore/fleece/MArray.java

```diff
@@ -13,98 +13,189 @@
-    }
 
-    public boolean set(int i, Object val) {
-        return set(_handle, i, val);
+    public FLArray getBaseArray() {
+        return _baseArray;
     }
 
-    public boolean insert(int i, Object val) {
-        return insert(_handle, i, val);
-    }
+    /* Public Methods */
 
-    public boolean append(Object val) {
-        return append(_handle, val);
-    }
+    /** Returns a reference to the MValue of the item at the given index.
+     If the index is out of range, returns an empty MValue. */
+    public MValue get(long index) {
+        if (index < 0 || index >= _values.size())
+            return MValue.EMPTY;
 
-    public boolean remove(int i) {
-        return remove(_handle, i, 1);
+        MValue value = _values.get((int)index);
+        if (value.isEmpty()) {
+            if (_baseArray != null) {
+                value = new MValue(_baseArray.get(index));
+                _values.set((int) index, value);
+            }
+        }
+        return value;
     }
 
-    public boolean remove(int i, int n) {
-        return remove(_handle, i, n);
-    }
+    public boolean set(long index, Object value) {
+        if (!isMutable())
+            throw new IllegalStateException("Cannot set items in a non-mutable MArray");
 
-    public boolean clear() {
-        return clear(_handle);
-    }
+        if (index < 0 || index >= count())
+            return false;
 
-    public void encodeTo(Encoder enc) {
-        encodeTo(_handle, enc._handle);
+        mutate();
+        _values.set((int)index, new MValue(value));
+        return true;
     }
 
-    @Override
-    protected void finalize() throws Throwable {
-        free();
-        super.finalize();
+    public boolean insert(long index, Object value) {
+        if (!isMutable())
+            throw new IllegalStateException("Cannot insert items in a non-mutable MArray");
+
+        if (index < 0 || index > count())
+            return false;
+
+        if (index < count())
+            populateValues();
+
+        mutate();
+        _values.add((int) index, new MValue(value));
+        return true;
     }
 
-    @Override
-    public void free() {
-        if (_handle != 0L && !_managed) {
-            free(_handle);
-            _handle = 0L;
-        }
+    public boolean append(Object value) {
+        return insert(count(), value);
     }
 
-    private static native void free(long handle);
+    public boolean remove(long start, long num) {
+        if (!isMutable())
+            throw new IllegalStateException("Cannot remove items in a non-mutable MArray");
 
-    private static native long init();
+        long end = start + num;
+        if (end <= start)
+            return end == start;
 
-    private static native void initInSlot(long handle, long mv, long parent);
+        long count = count();
+        if (end > count)
+            return false;
 
-    private static native void initAsCopyOf(long handle, long mDict, boolean isMutable);
+        if (end < count)
```

Comment:
Is calling `populateValues()` once enough? I think it is not necessary to call the method always if `end < count()`.

Reply:
I believe it should be done once, but I followed the original MArray's implementation. Can you create a ticket to revise the implementation?

https://github.com/couchbaselabs/fleece/blob/260479eac5b274fae9642135978ab7b7d6edd3c5/Fleece/Mutable/MArray.hh#L114-L115

Reply:
https://github.com/couchbaselabs/fleece/issues/27